src/core/manager.h

```diff
@@ -474,7 +474,15 @@ void manager_clear_jobs(Manager *m);
 
 void manager_unwatch_pid(Manager *m, pid_t pid);
 
-unsigned manager_dispatch_load_queue(Manager *m);
+unsigned int manager_dispatch_load_queue_realm(Manager *m);
+
+#define manager_dispatch_load_queue(m)                                  \
+        ({                                                              \
+                assert(m);                                              \
+                (!m->dispatching_load_queue)                            \
+                        ? manager_dispatch_load_queue_realm(m)          \
```

Comment:
I don't see why this would make such a big change. manager_dispatch_load_queue_realm() does a check for m->dispatching_load_queue as the very first thing. So the only thing this patch does is "inline" the check from the called function into the caller.

Reply:
Oh, thanks for your comments.
We have added some printing information to the code, so the improvement speed mentioned in the commit message is not universal, and we will make some corrections later.
Since this is the critical path of the code, we should probably inline it to remove unnecessary function call overhead; and we also observe that all functions that call manager_dispatch_load_queue () have already asserted that m is not NULL, so there is no need to assert it again.
We will resubmit a patch later.

Reply:
> we also observe that all functions that call manager_dispatch_load_queue () have already asserted that m is not NULL, so there is no need to assert it again.
We will resubmit a patch later.

Please no, if you're in such a performance-sensitive use case just produce a build without any asserts at all; add -DNDEBUG to your CFLAGS.

The asserts are there to codify programmer assumptions on a per-function basis.  If the function assumes it can dereference an input parameter, we assert that it's not NULL.

It's great that you're looking at optimizations, but at least start with an optimized build and use profiling tools to guide your efforts.

Reply:
Okay thank you.